### PR TITLE
VBIOS patching via ForceResolution option

### PR DESCRIPTION
**Pull request by Goldfish64**
https://github.com/acidanthera/bugtracker/issues/1205

Supports Intel graphics only.

### CODE REVIEW

File: Legacy/BootPlatform/BiosVideo/BiosVideo.c [hunk 1]
```diff
@@ -1068,14 +911,340 @@ BiosVideoCheckForVbe (
...
+
+  @param BiosVideoPrivate   Video child device private data structure
+
+**/
+VOID
+BiosVideoDeviceReleaseResource (
+  BIOS_VIDEO_DEV  *BiosVideoPrivate
+  )
+{
+  if (BiosVideoPrivate == NULL) {
```
Comment (vit9696) [inline Legacy/BootPlatform/BiosVideo/BiosVideo.c:1057]: Is the parameter optional?

File: Legacy/BootPlatform/BiosVideo/BiosVideo.c [hunk 2]
```diff
@@ -1068,14 +911,340 @@ BiosVideoCheckForVbe (
...
+  }
+
+  if (BiosVideoPrivate->DevicePath != NULL) {
+    gBS->FreePool (BiosVideoPrivate->DevicePath);
+  }
+
+  gBS->FreePool (BiosVideoPrivate);
+}
+
+#define PCI_DEVICE_ENABLED  (EFI_PCI_COMMAND_IO_SPACE | EFI_PCI_COMMAND_MEMORY_SPACE)
```
Comment (vit9696) [inline Legacy/BootPlatform/BiosVideo/BiosVideo.c:1129]: Was here before, but could we move it to a separate header?

File: Legacy/BootPlatform/BiosVideo/BiosVideo.c [hunk 3]
```diff
@@ -1068,14 +911,340 @@ BiosVideoCheckForVbe (
...
+  //
+  if ((Pci.Hdr.Command & PCI_DEVICE_ENABLED) == PCI_DEVICE_ENABLED) {
+    if (Pci.Hdr.ClassCode[2] == PCI_CLASS_OLD && Pci.Hdr.ClassCode[1] == PCI_CLASS_OLD_VGA) {
+      //
+      // Base Class 0x00 Sub-Class 0x01 - Backward compatible VGA device
+      //
+      VgaCompatible = TRUE;
+    }
+
+    if (Pci.Hdr.ClassCode[2] == PCI_CLASS_DISPLAY && Pci.Hdr.ClassCode[1] == PCI_CLASS_DISPLAY_VGA && Pci.Hdr.ClassCode[0] == 0x00) {
```
Comment (vit9696) [inline Legacy/BootPlatform/BiosVideo/BiosVideo.c:1176]: else if?

File: Legacy/BootPlatform/BiosVideo/BiosVideo.h [hunk 4]
```diff
@@ -572,4 +604,25 @@ LegacyBiosInt86 (
...
+
+  @retval EFI_SUCCESS         Resolution successfully forced.
+  @retval EFI_UNSUPPORTED     The video adapter or the display are not supported.
+  @retval EFI_ALREADY_STARTED The specified resolution is already supported.
+**/
+EFI_STATUS
+EFIAPI
+BiosVideoForceResolutionSetResolution (
+  IN OUT OC_FORCE_RESOLUTION_PROTOCOL *This,
+  IN     UINT32                       Width,
```
Comment (vit9696) [inline Legacy/BootPlatform/BiosVideo/BiosVideo.h:624]: Should probably be marked as OPTIONAL for Max.

File: Legacy/BootPlatform/BiosVideo/VideoBiosPatch.c [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+{
+  UINT32 Tmp1;
+  UINT32 Tmp2;
+  UINT32 Tmp3;
+
+  UINT32 Vbl;
+  UINT32 VFreq;
+  UINT32 Hbl;
+
+  Tmp1  = 11 * Frequency;
```
Comment (vit9696) [inline Legacy/BootPlatform/BiosVideo/VideoBiosPatch.c:100]: Could you add any reference to an algorithm on how these values are calculated. Like a spec or a driver.

File: Legacy/BootPlatform/BiosVideo/VideoBiosPatch.c [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+  UINT16      VSyncEnd;
+  UINT16      VBlank;
+
+  EDID_DTD    Dtd;
+
+  //
+  // DTD entry for 1024x768 resolution for the Intel VBIOS.
+  // This will be replaced with our custom resolution.
+  //
+  CONST UINT8 IntelDtd1024[] = {
```
Comment (vit9696) [inline Legacy/BootPlatform/BiosVideo/VideoBiosPatch.c:196]: This wants to have STATIC.

File: OpenDuetPkg.dsc [hunk 7]
```diff
@@ -97,11 +97,18 @@
   HobLib|MdePkg/Library/DxeHobLib/DxeHobLib.inf
   CpuExceptionHandlerLib|MdeModulePkg/Library/CpuExceptionHandlerLibNull/CpuExceptionHandlerLibNull.inf
   ExtractGuidedSectionLib|MdePkg/Library/DxeExtractGuidedSectionLib/DxeExtractGuidedSectionLib.inf
+  OcDebugLogLib|OpenCorePkg/Library/OcDebugLogLibNull/OcDebugLogLibNull.inf
```
Comment (vit9696) [inline OpenDuetPkg.dsc:100]: Do not think we need OcDebugLogLib. Just DebugLib will be enough throughout the code.

File: Library/OcDevicePathLib/OcDevicePathLib.inf [hunk 8]
```diff
@@ -44,6 +44,7 @@
   DebugLib
   DevicePathLib
   MemoryAllocationLib
+  OcDebugLogLib
```
Comment (vit9696) [inline Library/OcDevicePathLib/OcDevicePathLib.inf:47]: Better just use DebugLib.

**Comment by vit9696** [comment]
The changes look very good. I marked a few places it would be good to update before merging. Also, could you upload the binaries so that I could let some people test them here with legacy setups for (e.g. performance) regressions.

File: Docs/Configuration.tex [hunk 9]
```diff
@@ -5966,6 +5966,19 @@ \subsection{Output Properties}\label{uefioutputprops}
...
+  \textbf{Type}: \texttt{plist\ boolean}\\
+  \textbf{Failsafe}: \texttt{false}\\
+  \textbf{Description}: Forces \texttt{Resolution} to be set in cases where the desired
+  resolution is not available by default, such as on legacy Intel GMA and first
+  generation Intel HD Graphics (Ironlake/Arrandale). Setting \texttt{Resolution} to
+  \texttt{Max} will try to pull the largest available resolution from the connected
+  display's EDID.
+
+  \emph{Note}: This option depends on the \href{https://github.com/acidanthera/OpenCorePkg/blob/master/Include/Acidanthera/Protocol/OcForceResolution.h}{\texttt{OC\_FORCE\_RESOLUTION\_PROTOCOL}}
+  protocol being present. This protocol is currently only supported by \texttt{OpenDuetPkg}.
```
Comment (vit9696) [inline Docs/Configuration.tex:5980]: Should probably add for Intel GPUs since in theory at least AMD GPUs may also need that.

**Comment by Sniki** [comment]
@vit9696 @Goldfish64 
Does this PR fix the garbled screen problem on some devices firmware if CSM is disabled

Usually ThinkPads Laptops like Haswell Series have garbled screen if CSM is disabled, the only known solution at that time was to turn CSM On but with CSM turned On the problem is that we cannot enable Secure Boot.

**Comment by Andrey1970AppleLife** [comment]
> @vit9696 @Goldfish64
> Does this PR fix the garbled screen problem on some devices firmware if CSM is disabled when running macOS
> 
> Usually ThinkPads Laptops like Haswell Series have garbled screen if CSM is disabled, the only known solution at that time was to turn CSM On but with CSM turned On the problem is that we cannot enable Secure Boot.

Method 4.
https://applelife.ru/threads/acer-aspire-serij-v3-v5-v7-na-processorax-haswell.41473/page-76#post-874880

**Comment by Sniki** [comment]
> > @vit9696 @Goldfish64
> > Does this PR fix the garbled screen problem on some devices firmware if CSM is disabled when running macOS
> > Usually ThinkPads Laptops like Haswell Series have garbled screen if CSM is disabled, the only known solution at that time was to turn CSM On but with CSM turned On the problem is that we cannot enable Secure Boot.
> 
> Method 4.
> https://applelife.ru/threads/acer-aspire-serij-v3-v5-v7-na-processorax-haswell.41473/page-76#post-874880

Thanks man, worked like charm !

Question, would it be possible to add a device property like 'HorizontalSyncPulseWidth' to DeviceProperties into IGPU so there is no need for everyone to individually patch EDID to fix this issue.

**Comment by MacNB** [comment]
I know this closed but thought I ask anyway.
Were there any plans to enhance **ForceResolution** to support Nvidia GPU's ??
Scenario:
GPU: GT730 (Kepler GK208B) on a Legacy (Duet) Dell Xeon system
Recently upgraded the Dell Monitor from FHD to UHD (S2721QS) and now OpenCanopy can no longer display GUI in FHD (1920x1080) but Max 1280x1024 which looks stretched. And no 3840x2160. Log shows:
`00:321 00:002 OCAE: Set screen resolution to 800x600 - Success`

later in the log:
`00:902 00:077 OCC: GOP exists on ConsoleOutHandle and has 4 modes`
`00:904 00:002 OC: Requested resolution is 0x0@0 (max: 1, force: 0) from Max`
`00:906 00:002 OCC: Requesting 0x0@0 (max: 1) resolution, curr 1, total 4`
`00:908 00:002 OCC: Current FB at 0xD7000000 (0xE00000), format 1, res 800x600 scan 800`
`00:911 00:002 OCC: Mode 0 - 640x480:1`
`00:913 00:002 OCC: Mode 1 - 800x600:1`
`00:915 00:002 OCC: Mode 2 - 1024x768:1`
`00:917 00:002 OCC: Mode 3 - 1280x1024:1`
`00:919 00:002 OCC: Setting mode 3 with 1280x1024 resolution`
`01:173 00:253 OCC: Changed resolution mode to 3`
`01:175 00:002 OC: Changed resolution to 0x0@0 (max: 1, force: 0) from Max - Success`
`01:177 00:002 OC: Setting UIScale to 1 - Success`
`01:180 00:002 OCC: Using builtin text renderer with 1 scale`

Once booted in macOS, all is fine with the resolution (3840x2160 displayed as 1920x1080 retina)

Ideally, it would great to set `ForceResolution` to 3840x2160 and set UIScale to 2.
Or, even just 1920x1080

**Pull request merged**